### PR TITLE
Fixed acyclic coloring

### LINKED ISSUE #75: acyclic coloring test error
@pkj-m could you take a look into https://travis-ci.org/JuliaDiffEq/SparseDiffTools.jl/jobs/615131888#L323-L352 ?

### PR DESCRIPTION
**Pull request by pkj-m**
Implemented modified version of acyclic coloring algorithm using disjoint sets with root as integer (representing the edge's position in the graph) instead of the edge object itself to avoid unforeseen errors. Also enforced initialization of the tuple vectors with `(0,0)` instead of `undef` to avoid random errors.

Few performance enhancements that are possible in this implementation:
1. use hashmap to save the edges' keys instead of iterating every time to find edge index
2. implement tests using the standard `SimpleGraphs(nv, ne)` function instead of generating graphs using `rand(1:nv)`

Fixes #75

### CODE REVIEW

File: src/coloring/acyclic_coloring.jl [hunk 1]
```diff
@@ -10,29 +10,30 @@ is a collection of trees—and hence is acyclic.
 Reference: Gebremedhin AH, Manne F, Pothen A. **New Acyclic and Star Coloring Algorithms with Application to Computing Hessians**
 """
 function color_graph(g::LightGraphs.AbstractGraph, ::AcyclicColoring)
-
     color = zeros(Int, nv(g))
-    forbidden_colors = zeros(Int, nv(g))
+    set = DisjointSets{Int}([])
+
+    first_visit_to_tree = Array{Tuple{Int, Int}, 1}()
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: We can switch to `Vector{T}` everywhere in the code base

File: src/coloring/acyclic_coloring.jl [hunk 2]
```diff
@@ -10,29 +10,30 @@ is a collection of trees—and hence is acyclic.
...
+    first_neighbor = Array{Tuple{Int, Int}, 1}()
 
-    set = DisjointSets{LightGraphs.Edge}([])
+    init_array!(first_visit_to_tree, ne(g))
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: you can replace the variable initialization and filling with:
```julia
fill((0,0), ne(g))
```

File: src/coloring/acyclic_coloring.jl [hunk 3]
```diff
@@ -10,29 +10,30 @@ is a collection of trees—and hence is acyclic.
...
-            if color[w] != 0
+            if color[w]!=0
                 forbidden_colors[color[w]] = v
             end
         end
-        #enforces the second condition of acyclic coloring
+
         for w in outneighbors(g, v)
-            if color[w] != 0 #colored neighbor
+            if color[w]!=0
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: Keep spaces around operators

```suggestion
            if color[w] != 0
```

File: src/coloring/acyclic_coloring.jl [hunk 4]
```diff
@@ -10,29 +10,30 @@ is a collection of trees—and hence is acyclic.
 Reference: Gebremedhin AH, Manne F, Pothen A. **New Acyclic and Star Coloring Algorithms with Application to Computing Hessians**
 """
 function color_graph(g::LightGraphs.AbstractGraph, ::AcyclicColoring)
-
     color = zeros(Int, nv(g))
-    forbidden_colors = zeros(Int, nv(g))
+    set = DisjointSets{Int}([])
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: Give `set` a more informative name

```suggestion
    set = DisjointSets{Int}(())
```

File: src/coloring/acyclic_coloring.jl [hunk 5]
```diff
@@ -41,30 +42,28 @@ function color_graph(g::LightGraphs.AbstractGraph, ::AcyclicColoring)
...
             end
         end
     end
-
     return color
 end
 
+
 """
     prevent_cycle(v::Integer,
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: Update the docstring for `preven_cycle`, add the "!" and check the function signature

File: src/coloring/acyclic_coloring.jl [hunk 6]
```diff
@@ -120,25 +110,22 @@ previously uncolored vertex v, by comparing it with the adjacent vertex w.
...
 """
-function grow_star!(set::DisjointSets{LightGraphs.Edge},
-                   v::Integer,
-                   w::Integer,
-                   g::LightGraphs.AbstractGraph,
-                   first_neighbor::AbstractVector{<:Tuple{Integer, Integer}},
-                   color::AbstractVector{<: Integer})
-    edge = find_edge(g, v, w)
-    push!(set, edge)
+function grow_star!(v::Integer,
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: As a convention, the mutated argument(s) goes first in the argument list

File: src/coloring/acyclic_coloring.jl [hunk 7]
```diff
@@ -120,25 +110,22 @@ previously uncolored vertex v, by comparing it with the adjacent vertex w.
...
+                    w::Integer,
+                    g::LightGraphs.AbstractGraph,
+                    set::DisjointSets{<:Integer},
+                    first_neighbor::Array{<: Tuple{Integer,Integer},1},
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]:
```suggestion
                    first_neighbor::Vector{<: Tuple{Integer,Integer}},
```

File: src/coloring/acyclic_coloring.jl [hunk 8]
```diff
@@ -152,51 +139,93 @@ end
...
+"""
+        make_set!(v::Integer,
+                  w::Integer,
+                  g::LightGraphs.AbstractGraph,
+                  set::DisjointSets{<:Integer})
+
+creates a new singleton set in the disjoint set 'set' consisting
+of the edge connecting v and w in the graph g
+"""
+function make_set!(v::Integer,
```
Comment (matbesancon) [inline src/coloring/acyclic_coloring.jl]: This function name should have a more explicit name. `make_set!` sounds like it builds a new set object, not mutates a disjoint set structure

File: test/test_acyclic.jl [hunk 9]
```diff
@@ -3,8 +3,9 @@ using LightGraphs
 using Test
 
 using Random
-Random.seed!(123)
+Random.seed!(45)
```
Comment (matbesancon) [inline test/test_acyclic.jl]: Why changing the seed?

Comment (pkj-m) [reply test/test_acyclic.jl]: Oh, I was trying out different inputs for the algorithm by changing the seed at every run. Forgot to set it back to the original value.

**Comment by matbesancon** [comment]
Still some errors, do you know where to look at?

**Comment by pkj-m** [comment]
@matbesancon are you talking about errors in this PR or the travis build? It seems the build is failing because of some issues with methods related to BlockBandedMatrices. Acyclic tests are passing.

**Comment by matbesancon** [comment]
Dang thanks, yes I was talking about the Travis build

**Pull request merged**